Add TagsBox widget for React IS (#10)
algolia/instantsearch-labs: tagsbox/react-instantsearch/README.md

# TagsBox Widget

🎥  **[See it Live on CodeSandbox](https://codesandbox.io/s/vyvy30vyy)**

## Demo

![Demo of TagsBox Widget](https://cl.ly/6ea705021fda/Screen%252520Recording%2525202018-10-26%252520at%25252001.35%252520PM.gif)

## Description

> This is the `TagsBox` Labs widget for [InstantSearch.js](https://community.algolia.com/instantsearch.js/). You can use this widget to search and select tags with Algolia and use them later on for any purpose want.

Author: Sorry about this one, will have to remember it!

> This is the `TagsBox` Labs widget for [InstantSearch.js](https://community.algolia.com/instantsearch.js/). You can use this widget to search and select tags with Algolia and use them later on for any purpose want.

Reviewer: no problemo!

This widget displays an input that will offer tag suggestions while user is typing. Tags can be then clicked to be added to the input.

This helps user to select records for another purpose than searching through them but to reuse them to complete another action later (like choosing a contact for sending an email or a category while creating a blog post for instance).

**Table of Contents**

* [Get the code](#get-the-code)
  * [JavaScript](#javascript)
  * [CSS](#css)
* [Usage](#usage)
  * [Requirements](#requirements)
* [API](#api)
  * [relevanceWidget(opts)](#relevancewidgetopts)
* [Examples](#examples)
* [Implementation details](#implementation-details)
* [Contributing](#contributing)

## Get the code

This widget comes with JavaScript but also pre-defined CSS.

### JavaScript

You can copy and paste the JavaScript code from the repository itself, grab it in the [src/lib](src/lib) folder and use the `<Tag />` component.

### CSS

You can copy and paste the necessary CSS code from the repository itself, grab it in [src/lib/Tags.css](src/lib/Tags.css).

## Usage

The simplest usage is:

```jsx
import { InstantSearch } from 'react-instantsearch-dom';

> import TagsBox from 'instansearch-tagsbox-react';

Reviewer 2: For latter lab experiments, should we create a grouped package or specific packages like this one?

Author: That's a really good one :) I'm certainly not the good person to answer here but will comply once a decision has been made!

export const TagSelectedComponent = ({ hit }) => (
    <Fragment>
        {hit.lastname}
    </Fragment>
);

export const TagSuggestionComponent = ({ hit }) => (
    <Fragment>
        {hit.firstname} {hit.lastname}
    </Fragment>
);

<InstantSearch
    appId="..."
    apiKey="..."
    indexName="..."
>
    <TagsBox
        tagSelectedComponent={TagSelectedComponent}
        tagSuggestionComponent={TagSuggestionComponent}

>         onUpdate={(newTags, oldTags) => console.log}

Reviewer 2: The `console.log` wouldn't work like that. I would rename `oldTags` to `previousTags`.

```suggestion
        onUpdate={(newTags, previousTags) => console.log(newTags, previousTags)}
```

        translations={{ placeholder: "Search…", noResult: "…" }}
    />
</InstantSearch>
```

### Requirements

> You should encapsulte the `<TagsBox />` component inside an `<InstantSearch />` one.

Reviewer 2: ```suggestion
You should encapsulate the `<TagsBox />` component inside an `<InstantSearch />` one.
```

To work properly, **you'll need to provide three props to the component**:

- `tagSelectedComponent` (component): describe how a selected tag should be displayed
  - receive a `hit` as parameter (coming from Algolia)
- `tagSuggestionComponent` (component): describe how a suggested tag should be displayed
  - receive a `hit` as parameter (coming from Algolia)

> - `onUpdate` (function): call each time a tag is added or removed

Reviewer 2: ```suggestion
- `onUpdate` (function): call each time a tag is added or removed
```

>   - receive `actualTags` and `oldTags` as parameters, both arrays that respectively contains the current selected tags and the previous ones

Reviewer 2: ```suggestion
  - receive `newTags` and `previousTags` as parameters, both arrays that respectively contains the current selected tags and the previous ones
```

> It is possible to get encapsulate the `<TagBox />` component in an `<Index />` one if you wish to target different indices from the same InstantSearch instance.

Reviewer 2: ```suggestion
It is possible to encapsulate the `<TagBox />` component in an `<Index />` one if you wish to target different indices from the same InstantSearch instance.
```

## Examples

* using the `connectAutoComplete()` connector: [CodeSandbox](https://codesandbox.io/s/vyvy30vyy)

### Props

| Name                              | Type      | Use                                                 |
| --------------------------------- | --------- | --------------------------------------------------- |

> | tagSelectedComponent (required)   | Component | Describe how a selected tag should be displayed.    |

Reviewer 2: I would probably call this `selectedTagComponent`.

> | tagSuggestionComponent (required) | Component | Describe how a suggested tag should be displayed.   |

Reviewer 2: I would call this `suggestedTagComponent`.

| onUpdate (required)               | Function  | Call each time a tag is added or removed.           |

> | createTagComponent                | Component | Describe how an unexisting tag should be displayed. |

Reviewer 2: Not sure about this suggestion but maybe call it something like `noResultsComponent`?

I feel like it would make sense to expose the internal `onTagAdd` and to let the user use it on click on this `noResultsComponent` so that this component is not exclusively used to create a new tag. It would remove the props `createTagAttribute`.

## Description

> This is the `TagsBox` Labs widget for [React InstantSearch](https://community.algolia.com/instantsearch.js/). You can use this widget to search and select tags with Algolia and use them later on for any purpose want.

Reviewer: the link as well :)

Reviewer: ```suggestion
This is the `TagsBox` Labs widget for [React InstantSearch](https://community.algolia.com/react-instantsearch/). You can use this widget to search and select tags with Algolia and use them later on for any purpose want.
```

Author: 🤭
